Gpytorch (#586)
argonne-lcf/user-guides: docs/polaris/data-science/frameworks/gpytorch.md

# Guide of Login to Polaris Load Environments, Run Jobs, and Install GPytorch

> #### 1. Login and queue a job

Reviewer: Why do you go from # section heading to #### ?

Author: The first one is the doc title and the second a section. If it is weird, I can change it.

Reviewer: Yes, I understand that, but why not use `##` for the section instead of `####`? I see the same issue in the new PR #635

Login to Polaris
```
ssh alcfusername@polaris.alcf.anl.gov
```

Start a sessiom, for example Interactive ssh Session on a Compute Node
```
qsub -I -l nodes=1:ppn=4 -l walltime=1:00:00 -q debug -l filesystems=eagle:home -A datascience
```
  where:

> * -A for project name

Reviewer: or actually, delete this whole section since it is redundant with the pages on job submissionn. I woudl suggested replacing it with something similar to what is used at the top of 
https://docs.alcf.anl.gov/polaris/data-science/frameworks/deepspeed/

```
!!! note

    The instructions below should be **ran directly from a compute node**.

    Explicitly, to request an interactive job (from `polaris-login`):
    ```bash
    qsub -A <project> -q debug-scaling -l select=2 -l walltime=01:00:00 -I
    ```

    Refer to [job scheduling and
    execution](../../../running-jobs/job-and-queue-scheduling.md) for
    additional information.
```

* -q for single-gpu or full-node. If using full-node, it would be better to add `-M <youremail>` as well, so that you will receive an email when your job is starting.
* -n number of resources (n gpu or n node)
* -t number of minutes you want to have
* -I indicates an interactive session. One can also remove -I and specify a executable bash script for it to run directly on the compute node

If queueing for an interactive session, once it is running, we can use `qstat -u <yourusername>` **on a service node**  to see our job id and allocated node. 

> #### 2. Once on a Compute Node, Load Modules

Reviewer: ```suggestion
#### 2. Load modules
```

> Load the Conda Environment (Module) with Pytorch, since our GPytorch has Pytorch dependency

Reviewer: ```suggestion
Load the Anaconda environment module, which contains a PyTorch installation, since GPyTorch has PyTorch as a dependency:
```

```
module load conda
conda activate
```
* Notice that we can check the available modules with "module avail" and check the loaded modules with "module list"

Create a virtual environment with python and activate it 
```
python -m venv --system-site-packages path_to_myenv
source path_to_myenv/bin/activate
```
Now the bash prompt should show that we're in the environment we just created, and we're good to use pip install
```
pip install gpytorch==version
```

> ### After the First Time

Reviewer: ```suggestion
### Loading environment in future sessions
```

After the first time, to run the files, simply activate the python_venv on a compute node with 
```
module load conda
source path_to_myenv/bin/activate
``` 

> ## Using Jupyter Notebook to Run GPytorch on Polaris

Reviewer: ```suggestion
## Using Jupyter Notebook to run GPyTorch on Polaris
```

Here is the guide:

> ### Approach 1 - Use Jupyter Hub

Reviewer: ```suggestion
### Approach 1 - Use ALCF JupyterHub
```

and fix formatting below similarly

1. Go to [Jupyter Hub of ALCF](https://jupyter.alcf.anl.gov/), click Login Polaris
2. Queue up on a debug node

**For the first time only, one needs to set up the environment and Kernel by follow these extra steps**

4. Once Jupyter Notebook is launched on a compute node, click "New" and open a **terminal**
5. Run 
``` 
module load conda
conda activate
source <path_to_previously_created_python_venv>/bin/activate
python -m ipykernel install --user --name python_venv
```

> Note: depending on the system and environment, you might need to install the "ipykernel" package first. The python_venv that I just created has the ipykernel module.

Reviewer: ```suggestion
Note: depending on the system and environment, you might need to install the "ipykernel" package first. The `python_venv` that I just created has the `ipykernel` module.
```

> Go back to your .ipynb file, change kernel to python_venv from the dropdown menu, and we'll be good to run GPytorch!

Reviewer: ```suggestion
Go back to your `.ipynb` file, change kernel to `python_venv` from the dropdown menu, and we'll be good to run GPyTorch!
```

Reviewer: similarly below
